### ISSUE TITLE
org.h2.jdbc.JdbcSQLException: Feature not supported: "isWrapperFor"

### ISSUE THREAD

**Issue by dspinellis** [rundeck/rundeck#1175]
Rundeck version 2.5.0-1 installed as a package on [AWS Linux 2015.03](https://aws.amazon.com/amazon-linux-ami/2015.03-release-notes/) fills `/var/lib/rundeck/data/rundeckdb.trace.db` with stack traces of the exception `org.h2.jdbc.JdbcSQLException: Feature not supported: "isWrapperFor"`. [According to the H2 developers](http://h2-database.66688.n3.nabble.com/Feature-not-supported-quot-isWrapperFor-quot-50100-181-td4029763.html) this is a known issue, which was fixed in August 2014.  Consider updating the distributed H2 jar.

Here is the complete stack trace.
```
04-19 07:26:21 jdbc[3]: exception
org.h2.jdbc.JdbcSQLException: Feature not supported: "isWrapperFor" [50100-176]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:344)
        at org.h2.message.DbException.get(DbException.java:178)
        at org.h2.message.DbException.get(DbException.java:154)
        at org.h2.message.DbException.getUnsupportedException(DbException.java:215)
        at org.h2.message.TraceObject.unsupported(TraceObject.java:395)
        at org.h2.jdbc.JdbcStatement.isWrapperFor(JdbcStatement.java:1076)
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.isTypeOf(ResultSetReturnImpl.java:99)
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:70)
        at org.hibernate.loader.Loader.getResultSet(Loader.java:2065)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1862)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1838)
        at org.hibernate.loader.Loader.doQuery(Loader.java:909)
        at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:354)
        at org.hibernate.loader.Loader.doList(Loader.java:2553)
        at org.hibernate.loader.Loader.doList(Loader.java:2539)
        at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2369)
        at org.hibernate.loader.Loader.list(Loader.java:2364)
        at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:126)
        at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1682)
        at org.hibernate.internal.CriteriaImpl.list(CriteriaImpl.java:380)
        at org.codehaus.groovy.grails.orm.hibernate.metaclass.FindAllByPersistentMethod$1.doInHibernate(FindAllByPersistentMethod.java:114)
        at org.codehaus.groovy.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:179)
        at org.codehaus.groovy.grails.orm.hibernate.GrailsHibernateTemplate.executeFind(GrailsHibernateTemplate.java:127)
        at org.codehaus.groovy.grails.orm.hibernate.metaclass.FindAllByPersistentMethod.doInvokeInternalWithExpressions(FindAllByPersistentMethod.java:74)
        at org.codehaus.groovy.grails.orm.hibernate.metaclass.AbstractClausedStaticPersistentMethod.doInvokeInternal(AbstractClausedStaticPersistentMethod.java:544)
        at org.codehaus.groovy.grails.orm.hibernate.metaclass.AbstractClausedStaticPersistentMethod.doInvokeInternal(AbstractClausedStaticPersistentMethod.java:418)
        at org.codehaus.groovy.grails.orm.hibernate.metaclass.AbstractStaticPersistentMethod.invoke(AbstractStaticPersistentMethod.java:79)
        at org.codehaus.groovy.grails.orm.hibernate.metaclass.AbstractStaticPersistentMethod.invoke(AbstractStaticPersistentMethod.java:72)
        at org.codehaus.groovy.grails.commons.metaclass.StaticMethodInvocation$invoke$0.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:124)
        at org.grails.datastore.gorm.GormStaticApi.methodMissing(GormStaticApi.groovy:105)
        at rundeck.Execution.$static_methodMissing(Execution.groovy)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
        at groovy.lang.MetaClassImpl.invokeStaticMissingMethod(MetaClassImpl.java:1493)
        at groovy.lang.MetaClassImpl.invokeStaticMethod(MetaClassImpl.java:1481)
        at groovy.lang.ExpandoMetaClass.invokeStaticMethod(ExpandoMetaClass.java:1123)
        at org.codehaus.groovy.runtime.callsite.StaticMetaClassSite.call(StaticMetaClassSite.java:50)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:120)
        at rundeck.services.ExecutionService.cleanupRunningJobs(ExecutionService.groovy:562)
        at rundeck.services.ExecutionService$$FastClassBySpringCGLIB$$ffb73873.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:708)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:98)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:262)
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:95)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:644)
        at rundeck.services.ExecutionService$$EnhancerBySpringCGLIB$$ad7684a4.cleanupRunningJobs(<generated>)
        at rundeck.services.ExecutionService$cleanupRunningJobs.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
        at BootStrap$_closure1.doCall(BootStrap.groovy:265)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:324)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1207)
        at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1110)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1016)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1081)
        at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1110)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1016)
        at groovy.lang.Closure.call(Closure.java:423)
        at groovy.lang.Closure.call(Closure.java:417)
        at grails.util.Environment.evaluateEnvironmentSpecificBlock(Environment.java:327)
        at grails.util.Environment.executeForEnvironment(Environment.java:320)
        at grails.util.Environment.executeForCurrentEnvironment(Environment.java:296)
        at org.codehaus.groovy.grails.commons.DefaultGrailsBootstrapClass.callInit(DefaultGrailsBootstrapClass.java:60)
        at org.codehaus.groovy.grails.web.context.GrailsConfigUtils.executeGrailsBootstraps(GrailsConfigUtils.java:78)
        at org.codehaus.groovy.grails.web.context.GrailsContextLoaderListener.initWebApplicationContext(GrailsContextLoaderListener.java:110)
        at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:106)
        at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:733)
        at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:233)
        at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1214)
        at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:676)
        at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:455)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:59)
        at org.eclipse.jetty.server.handler.HandlerWrapper.doStart(HandlerWrapper.java:90)
        at org.eclipse.jetty.server.Server.doStart(Server.java:261)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:59)
        at com.dtolabs.rundeck.RunServer.run(RunServer.java:112)
        at com.dtolabs.rundeck.RunServer.main(RunServer.java:72)
```

**Comment by gschueler** [comment]
thanks, this seems to have started after updating hibernate to hibernate 4.  it might be possible to redirect/quell these messages, see http://h2database.com/html/features.html#other_logging

**Comment by gschueler** [comment]
this seems to be a workaround, and quells all of the h2database trace log output:

1. append `;TRACE_LEVEL_FILE=4` to the dataSource.url value in rundeck-config.properties. e.g.:

    ``` {.properties}
    dataSource.url = jdbc:h2:file:/var/lib/rundeck/data/rundeckdb;MVCC=true;TRACE_LEVEL_FILE=4
    ```

2. modify /var/lib/rundeck/exp/webapp/WEB-INF/classes/log4j.properties, and disable the `h2database` logger, by appending this:

    ``` {.properties}
    log4j.logger.h2database = off
    ```

(Or configure log4j to use a rolling file appender with a file size limit.)

**Comment by gschueler** [comment]
@dspinellis have you tried replacing the h2 jar in your install with a newer one?  I have built and run rundeck through our automated tests using the latest h2 lib

I'd like to upgrade the h2 lib, but I don't want to make that change in a 2.5.1 release. I think perhaps the best way for now is to add the workaround to the config settings for 2.5.1, and postpone the h2 lib upgrade until 2.6.

**Comment by dspinellis** [comment]
Thank you; I will will work with the workaround you posted until 2.6.

**Comment by dspinellis** [comment]
The workaround addresses the problem; thanks!

**Comment by akusbharat** [comment]
We still see this same issue in 2.6.2 version as well. We upgraded from 2.0.3 to 2.6.2 this weekend

**Comment by gschueler** [comment]
@akusbharat make sure you apply the workarounds, they are now the standard config for our release, we have not yet upgraded the underlying h2 lib

**Comment by akusbharat** [comment]
Adding ;TRACE_LEVEL_FILE=4 to dataSource.url in rundeck-config.properties seems to have fixed the issue in 2.6.2. 

Thanks @gschueler
